packages/block-editor/src/components/dimensions-tool/index.js

```diff
@@ -0,0 +1,182 @@
+/**
+ * WordPress dependencies
+ */
+import { __experimentalUseCustomUnits as useCustomUnits } from '@wordpress/components';
+
+/**
+ * Internal dependencies
+ */
+import useSetting from '../use-setting';
+import AspectRatioTool from './aspect-ratio-tool';
+import ScaleTool from './scale-tool';
+import WidthHeightTool from './width-height-tool';
+
+/**
+ * @typedef {import('@wordpress/components/build-types/select-control/types').SelectControlProps} SelectControlProps
+ */
+
+/**
+ * @typedef {import('@wordpress/components/build-types/unit-control/types').WPUnitControlUnit} WPUnitControlUnit
+ */
+
+/**
+ * @typedef {Object} Dimensions
+ * @property {number} [width]       CSS width property.
+ * @property {number} [height]      CSS height property.
+ * @property {string} [scale]       CSS object-fit property.
+ * @property {string} [aspectRatio] CSS aspect-ratio property.
+ */
+
+/**
+ * @callback DimensionsControlsOnChange
+ * @param {Dimensions} nextValue
+ * @return {void}
+ */
+
+/**
+ * @typedef {Object} DimensionsControlsProps
+ * @property {string}                     [panelId]            ID of the panel that contains the controls.
+ * @property {Dimensions}                 [value]              Current dimensions values.
+ * @property {DimensionsControlsOnChange} [onChange]           Callback to update the dimensions values.
+ * @property {SelectControlProps[]}       [aspectRatioOptions] Aspect ratio options.
+ * @property {SelectControlProps[]}       [scaleOptions]       Scale options.
+ * @property {WPUnitControlUnit[]}        [unitsOptions]       Units options.
+ */
+
+/**
+ * Component that renders controls to edit the dimensions of an image or container.
+ *
+ * @param {DimensionsControlsProps} props The component props.
+ *
+ * @return {WPElement} The dimensions controls.
+ */
+function DimensionsControls( {
+	panelId,
+	value = {},
+	onChange = () => {},
+	aspectRatioOptions, // Default value is in AspectRatioTool.
+	scaleOptions, // Default value is in ScaleTool.
+	unitsOptions,
+} ) {
+	// TODO: 'spacing.units` was used in the featured image code, but we're
+	// setting width and height, not spacing here, so I'm not sure that's the
+	// right thing to use.
+	const customUnits = useCustomUnits( {
+		availableUnits: useSetting( 'spacing.units' ) || [
+			'px',
+			'%',
+			'vw',
+			'em',
+			'rem',
+		],
+	} );
+	const units = unitsOptions ?? customUnits;
+
+	// 'custom' is not a valid value for CSS aspect-ratio, but it is used in the
+	// dropdown to indicate that setting both the width and height is the same
+	// as a custom aspect ratio.
+	const aspectRatioValue =
+		value.width !== null &&
+		value.width !== undefined &&
+		value.height !== null &&
+		value.height !== undefined
```

Comment:
As these tests happen multiple times it might be worth extracting a well named `const`.